File: examples/python/inventory/handlers/release_reservation.py
"""Release reservation command handler."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains

from state import InventoryState
from handlers.errors import CommandRejectedError


def handle_release_reservation(command_book, command_any, state: InventoryState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Inventory not initialized")

    cmd = domains.ReleaseReservation()
    command_any.Unpack(cmd)

    if not cmd.order_id:
        raise CommandRejectedError("Order ID is required")
    if cmd.order_id not in state.reservations:
        raise CommandRejectedError("No reservation found for this order")

    qty = state.reservations[cmd.order_id]

    log.info("releasing_reservation", order_id=cmd.order_id, quantity=qty)

    event = domains.ReservationReleased(
        order_id=cmd.order_id,
        quantity=qty,
        new_available=state.available() + qty,
        released_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )

File: examples/go/inventory/logic/release_reservation.go
package logic

import (
	"angzarr"
	angzarrpb "angzarr/proto/angzarr"
	"inventory/proto/examples"

	goproto "google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// HandleReleaseReservation validates and creates a ReservationReleased event.
func HandleReleaseReservation(cb *angzarrpb.CommandBook, data []byte, state *InventoryState, seq uint32) (*angzarrpb.EventBook, error) {
	var cmd examples.ReleaseReservation
	if err := goproto.Unmarshal(data, &cmd); err != nil {
		return nil, angzarr.NewInvalidArgument("failed to unmarshal command: " + err.Error())
	}

	if !state.Exists() {
		return nil, angzarr.NewFailedPrecondition(ErrMsgNotInitialized)
	}
	if cmd.OrderId == "" {
		return nil, angzarr.NewInvalidArgument(ErrMsgOrderIDRequired)
	}
	qty, exists := state.Reservations[cmd.OrderId]
	if !exists {
		return nil, angzarr.NewFailedPrecondition(ErrMsgReservationNotFound)
	}

	return angzarr.PackEvent(cb.Cover, &examples.ReservationReleased{
		OrderId:      cmd.OrderId,
		Quantity:     qty,
		NewAvailable: state.Available() + qty,
		ReleasedAt:   timestamppb.Now(),
	}, seq)
}
